### PR TITLE
[netcore] Add libc name transition

### LINKED ISSUE #16521: [netcore] Implement DllMap
Run the following snippet on macOS:
```csharp
using System;
using System.Runtime.InteropServices;

class Program
{
    [DllImport("msvcrt")]
    public static extern int rand();

    static void Main(string[] args)
    {
        Console.WriteLine("Rand: " + rand());
    }
}
```

legacy mono will redirect "msvcrt" to "libc" (see https://github.com/mono/mono/blob/master/data/config.in#L18) and return some random value,

Mono-netcore:
```
Unhandled Exception:
System.DllNotFoundException: msvcrt assembly:<unknown assembly> type:<unknown type> member:(null)
   at HelloWorld.Program.Main(String[] args)
[ERROR] FATAL UNHANDLED EXCEPTION: System.DllNotFoundException: msvcrt assembly:<unknown assembly> type:<unknown type> member:(null)
   at HelloWorld.Program.Main(String[] args)
```

### PR DESCRIPTION
**Pull request by CoffeeFlux**
Fixes #16521, or at least as close as we're going to get for now.

### CODE REVIEW

File: mono/utils/mono-dl.c [hunk 1]
```diff
@@ -143,6 +148,24 @@ get_dl_name_from_libtool (const char *libtool_file)
 	return line;
 }
 
+static const char *
+fix_libc_name (const char *name)
```
Comment (vargaz) [inline mono/utils/mono-dl.c]: This should be inside an #ifdef ENABLE_NETCORE.

Comment (CoffeeFlux) [reply mono/utils/mono-dl.c]: I think the linker should get rid of it? Updated though, can't hurt.

**Comment by vargaz** [comment]
It will, but it will cause a warning.

File: mono/utils/mono-dl.c [hunk 2]
```diff
@@ -143,6 +148,26 @@ get_dl_name_from_libtool (const char *libtool_file)
...
 }
 
+#ifdef ENABLE_NETCORE
+static const char *
+fix_libc_name (const char *name)
+{
+	if (strncmp (name, "libc", 4) == 0) {
+		// Taken from CoreCLR: https://github.com/dotnet/coreclr/blob/6b0dab793260d36e35d66c82678c63046828d01b/src/pal/src/loader/module.cpp#L568-L576
+#if defined (HOST_DARWIN)
+		return "/usr/lib/libc.dylib";
```
Comment (lambdageek) [inline mono/utils/mono-dl.c]: A bit surprised this one hardcodes an absolute path.

Will this work on iOS?

File: mono/utils/mono-dl.c [hunk 3]
```diff
@@ -143,6 +148,26 @@ get_dl_name_from_libtool (const char *libtool_file)
...
+#elif defined (__FreeBSD__)
+		return "libc.so.7";
+#elif defined (LIBC_SO)
+		return LIBC_SO;
```
Comment (lambdageek) [inline mono/utils/mono-dl.c]: Once we have netcore on Android we should check that this define works there.

Comment (filipnavara) [reply mono/utils/mono-dl.c]: It does. The same code is used by CoreCLR and it was already compiled on Android few times.

File: mono/utils/mono-dl.c (hunk 2 above)
Comment (lewurm) [reply mono/utils/mono-dl.c]: **TL;DR**: It should be fine on iOS/tvOS/watchOS too.

On macOS:
```console
$ ls -l /usr/lib/libc.dylib
lrwxr-xr-x  1 root  wheel  15 Oct  1  2018 /usr/lib/libc.dylib -> libSystem.dylib
$ ls -l /usr/lib/libSystem.dylib
lrwxr-xr-x  1 root  wheel  17 Oct  1  2018 /usr/lib/libSystem.dylib -> libSystem.B.dylib
$ ls -l /usr/lib/libSystem.B.dylib
-rwxr-xr-x  1 root  wheel  60880 Aug 23 08:30 /usr/lib/libSystem.B.dylib
```
So the `libc.dylib` symlink ends up in `libSystem.B.dylib` which is suspiciously small to be honest.  All the symbols you would expect from `libc` seem to be in `/usr/lib/dyld`, which is the dynamic linker itself:
```console
$ file /usr/lib/dyld
/usr/lib/dyld: Mach-O universal binary with 2 architectures: [x86_64:Mach-O 64-bit dynamic linker x86_64] [i386:Mach-O dynamic linker i386]
/usr/lib/dyld (for architecture x86_64):        Mach-O 64-bit dynamic linker x86_64
/usr/lib/dyld (for architecture i386):  Mach-O dynamic linker i386
```
Let's add some verbose logging:
```console
$ DYLD_PRINT_LIBRARIES=y ls
dyld: /bin/ls
dyld: /usr/lib/libutil.dylib
dyld: /usr/lib/libncurses.5.4.dylib
dyld: /usr/lib/libSystem.B.dylib
[...]
```
No trace of `libc.dylib`.

So what's the situation on iOS? Not 100% sure, as I don't have direct shell access, there are however libs part of Xcode to support debugging:
```console
$ pwd
/Users/lewurm/Library/Developer/Xcode/iOS DeviceSupport/13.1.2 (17A860)/Symbols/usr/lib
$ ls -lha libc.dylib
ls: libc.dylib: No such file or directory
$ ls -lha libSystem.B.dylib
-rw-r--r--  1 lewurm  staff    12K Oct  2 10:13 libSystem.B.dylib
$ file dyld
dyld: Mach-O 64-bit dynamic linker arm64
```
My guess is that the iOS device will contain the symlinks as we have seen above on macOS. Let's snoop around with `lldb` on an iOS device:
```
(lldb) settings set target.env-vars DYLD_PRINT_LIBRARIES=y
(lldb) b xamarin_main
Breakpoint 1: where = sdbdebugtest`::xamarin_main(int, char **, XamarinLaunchMode) + 72 at monotouch-main.m:230:28, address = 0x0000000100412700
(lldb) r
Process 364 launched: '/Users/lewurm/Projects/sdbdebugtest/sdbdebugtest/bin/iPhone/Debug/device-builds/iphone9.3-13.1.2/sdbdebugtest.app/sdbdebugtest' (arm64)
dyld: <7D0702F8-6706-341E-AC5E-C51EA5735484> /private/var/containers/Bundle/Application/6B85FA1F-3D0F-4382-8EC2-4BC85661D704/sdbdebugtest.app/sdbdebugtest
dyld: <9B025821-A606-3634-A038-3032B850F8E2> /System/Library/Frameworks/UIKit.framework/UIKit
 <0DF2911E-80CB-3289-8A1E-ED0913D55A12> /System/Library/Frameworks/Foundation.framework/Foundation
dyld: <DD159500-7D53-3484-AA3D-9DD5073D7A2F> /System/Library/Frameworks/CFNetwork.framework/CFNetwork
Library/Frameworks/GSS.framework/GSS
dyld: <F8C60E32-C706-3653-AD1F-6BE0E1A67550> /usr/lib/libz.1.dylib
dyld: <C7159C1D-C5F2-3938-8019-5F1F26A812D6> /usr/lib/libiconv.2.dylib
dyld: <F11CB1CC-B616-39B1-8D76-192163AF9C98> /usr/lib/libSystem.B.dylib
[...]
```
Evidently it loads `libSystem.B.dylib` as well. Let's use `dlopen` directly:
```
(lldb) call (void *) dlopen("/usr/lib/libc.dylib", 0)
(void *) $0 = 0x000000000f490e00
(lldb) call (void *) dlopen("/usr/lib/libSystem.B.dylib", 0)
(void *) $2 = 0x000000000f490e00
(lldb) call (void *) dlopen("/usr/lib/libSystem.dylib", 0)
(void *) $3 = 0x000000000f490e00
(lldb) call (void *) dlopen("/usr/lib/libASDF.dylib", 0)
(void *) $4 = 0x0000000000000000
```
They all return the same address. Looks good to me!

File: mono/utils/mono-dl.c [hunk 4]
```diff
@@ -143,6 +148,26 @@ get_dl_name_from_libtool (const char *libtool_file)
 	return line;
 }
 
+#ifdef ENABLE_NETCORE
+static const char *
+fix_libc_name (const char *name)
+{
+	if (strncmp (name, "libc", 4) == 0) {
```
Comment (filipnavara) [inline mono/utils/mono-dl.c]: This will also trigger for `libcups`. Hence the unit test failures.

**Comment by filipnavara** [comment]
This broke my Mono LLVM performance runs:
```

	0x10defa8a8 - /Users/filipnavara/agent/_work/1/s/work/sdk-mono/shared/Microsoft.NETCore.App/5.0.0-alpha1.19456.10/libcoreclr.dylib : mono_dump_native_crash_info
	0x10de75525 - /Users/filipnavara/agent/_work/1/s/work/sdk-mono/shared/Microsoft.NETCore.App/5.0.0-alpha1.19456.10/libcoreclr.dylib : mono_handle_native_crash
	0x10def51a7 - /Users/filipnavara/agent/_work/1/s/work/sdk-mono/shared/Microsoft.NETCore.App/5.0.0-alpha1.19456.10/libcoreclr.dylib : altstack_handle_and_restore
	0x7fff7131f3b0 - /usr/lib/system/libsystem_platform.dylib : _platform_strcmp
	0x10e09d4bc - /Users/filipnavara/agent/_work/1/s/work/sdk-mono/shared/Microsoft.NETCore.App/5.0.0-alpha1.19456.10/libcoreclr.dylib : mono_dl_open
	0x10ded3352 - /Users/filipnavara/agent/_work/1/s/work/sdk-mono/shared/Microsoft.NETCore.App/5.0.0-alpha1.19456.10/libcoreclr.dylib : _ZZN11MonoLLVMJIT9addModuleEPN4llvm8FunctionENSt3__110shared_ptrINS0_6ModuleEEEENKUlRKNS3_12basic_stringIcNS3_11char_traitsIcEENS3_9allocatorIcEEEEE_clESE_
	0x10ded32a2 - /Users/filipnavara/agent/_work/1/s/work/sdk-mono/shared/Microsoft.NETCore.App/5.0.0-alpha1.19456.10/libcoreclr.dylib : _ZN4llvm3orc14LambdaResolverIZN11MonoLLVMJIT9addModuleEPNS_8FunctionENSt3__110shared_ptrINS_6ModuleEEEEUlRKNS5_12basic_stringIcNS5_11char_traitsIcEENS5_9allocatorIcEEEEE_ZNS2_9addModuleES4_S8_EUlSG_E0_E24findSymbolInLogicalDylibESG_
	0x10e0d57cf - /Users/filipnavara/agent/_work/1/s/work/sdk-mono/shared/Microsoft.NETCore.App/5.0.0-alpha1.19456.10/libcoreclr.dylib : _ZN4llvm15RuntimeDyldImpl22resolveExternalSymbolsEv
	0x10e0d5330 - /Users/filipnavara/agent/_work/1/s/work/sdk-mono/shared/Microsoft.NETCore.App/5.0.0-alpha1.19456.10/libcoreclr.dylib : _ZN4llvm15RuntimeDyldImpl18resolveRelocationsEv
	0x10e0d9d9e - /Users/filipnavara/agent/_work/1/s/work/sdk-mono/shared/Microsoft.NETCore.App/5.0.0-alpha1.19456.10/libcoreclr.dylib : _ZN4llvm11RuntimeDyld32finalizeWithMemoryManagerLockingEv
	0x10ded5b0e - /Users/filipnavara/agent/_work/1/s/work/sdk-mono/shared/Microsoft.NETCore.App/5.0.0-alpha1.19456.10/libcoreclr.dylib : _ZN4llvm3orc24RTDyldObjectLinkingLayer20ConcreteLinkedObjectINSt3__110shared_ptrINS_11RuntimeDyld13MemoryManagerEEENS4_INS_17JITSymbolResolverEEEZNS1_9addObjectENS4_INS_6object12OwningBinaryINSA_10ObjectFileEEEEES9_EUlNS3_15__list_iteratorINS3_10unique_pt
	0x10ded6438 - /Users/filipnavara/agent/_work/1/s/work/sdk-mono/shared/Microsoft.NETCore.App/5.0.0-alpha1.19456.10/libcoreclr.dylib : _ZZN4llvm3orc24RTDyldObjectLinkingLayer20ConcreteLinkedObjectINSt3__110shared_ptrINS_11RuntimeDyld13MemoryManagerEEENS4_INS_17JITSymbolResolverEEEZNS1_9addObjectENS4_INS_6object12OwningBinaryINSA_10ObjectFileEEEEES9_EUlNS3_15__list_iteratorINS3_10unique_p
	0x10ded63f2 - /Users/filipnavara/agent/_work/1/s/work/sdk-mono/shared/Microsoft.NETCore.App/5.0.0-alpha1.19456.10/libcoreclr.dylib : _ZNSt3__110__function6__funcIZN4llvm3orc24RTDyldObjectLinkingLayer20ConcreteLinkedObjectINS_10shared_ptrINS2_11RuntimeDyld13MemoryManagerEEENS6_INS2_17JITSymbolResolverEEEZNS4_9addObjectENS6_INS2_6object12OwningBinaryINSC_10ObjectFileEEEEESB_EUlNS_15__lis
	0x10ded269c - /Users/filipnavara/agent/_work/1/s/work/sdk-mono/shared/Microsoft.NETCore.App/5.0.0-alpha1.19456.10/libcoreclr.dylib : _ZN4llvm9JITSymbol10getAddressEv
	0x10ded0a4e - /Users/filipnavara/agent/_work/1/s/work/sdk-mono/shared/Microsoft.NETCore.App/5.0.0-alpha1.19456.10/libcoreclr.dylib : _ZN11MonoLLVMJIT7compileEPN4llvm8FunctionEiPP15LLVMOpaqueValuePPvS7_
	0x10deb31a1 - /Users/filipnavara/agent/_work/1/s/work/sdk-mono/shared/Microsoft.NETCore.App/5.0.0-alpha1.19456.10/libcoreclr.dylib : emit_method_inner
	0x10deb264c - /Users/filipnavara/agent/_work/1/s/work/sdk-mono/shared/Microsoft.NETCore.App/5.0.0-alpha1.19456.10/libcoreclr.dylib : mono_llvm_emit_method
	0x10ddccad7 - /Users/filipnavara/agent/_work/1/s/work/sdk-mono/shared/Microsoft.NETCore.App/5.0.0-alpha1.19456.10/libcoreclr.dylib : mini_method_compile
	0x10ddce31c - /Users/filipnavara/agent/_work/1/s/work/sdk-mono/shared/Microsoft.NETCore.App/5.0.0-alpha1.19456.10/libcoreclr.dylib : mono_jit_compile_method_inner
	0x10ddd20bf - /Users/filipnavara/agent/_work/1/s/work/sdk-mono/shared/Microsoft.NETCore.App/5.0.0-alpha1.19456.10/libcoreclr.dylib : mono_jit_compile_method_with_opt
	0x10de78292 - /Users/filipnavara/agent/_work/1/s/work/sdk-mono/shared/Microsoft.NETCore.App/5.0.0-alpha1.19456.10/libcoreclr.dylib : common_call_trampoline
	0x10de77cf0 - /Users/filipnavara/agent/_work/1/s/work/sdk-mono/shared/Microsoft.NETCore.App/5.0.0-alpha1.19456.10/libcoreclr.dylib : mono_magic_trampoline
```

**Comment by filipnavara** [comment]
The first parameter can be NULL (https://github.com/mono/mono/blob/5537a7c2ba6fca532c8b4e380278766d659981c0/mono/mini/llvm-jit.cpp#L172).

**Comment by akoeplinger** [comment]
Fixed by https://github.com/mono/mono/pull/17244

**Pull request merged**